core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -578,16 +584,22 @@ class LogManager(logDirs: Seq[File],
   }
 
   /**
-   * Make a checkpoint for all logs in provided directory.
-   */
-  private def checkpointLogRecoveryOffsetsInDir(dir: File): Unit = {
+    * Make a checkpoint for all logs in provided directory by default.
+    * The affectedLogs argument is an optimization that can be used if we only want to checkpoint a few partitions.
+    *
+    * @param dir the directory in which logs are checkpointed
+    * @param affectedLogs logs whose snapshots need to be cleaned
+    */
+  private def checkpointLogRecoveryOffsetsInDir(dir: File, affectedLogs: Option[Seq[Log]]): Unit = {
     for {
       partitionToLog <- logsByDir.get(dir.getAbsolutePath)
       checkpoint <- recoveryPointCheckpoints.get(dir)
     } {
       try {
         checkpoint.write(partitionToLog.mapValues(_.recoveryPoint))
-        allLogs.foreach(_.deleteSnapshotsAfterRecoveryPointCheckpoint())
+        affectedLogs.getOrElse(partitionToLog.values)
+          .filterNot(log => log.dir.getAbsolutePath.equals(dir.getAbsolutePath)) // it's safe to filter out the partitions with the incorrect directory
```

Comment:
It doesn't seem the above comment was addressed. log.dir.getAbsolutePath returns /disk1/topic-0 and dir.getAbsolutePath returns /disk1. So, we need to get the parent path of log (log.dir.getParent) to do the comparison. Also, it seems that we should use filter instead of filterNot? Could we make this method visible at the package level and add a unit test?

Also for the comment, perhaps change it to "to be safe, only keep partitions in the expected directory"?